File: scripts/clawtributor-campaign50/lanes/usage-known-zero-cost/verify-source.mjs
import assert from "node:assert/strict";
import { execFileSync } from "node:child_process";
import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const lane = path.dirname(fileURLToPath(import.meta.url));
const [target, phase] = process.argv.slice(2);
const spec = JSON.parse(readFileSync(path.join(lane, "source.json"), "utf8"));
assert.ok(["original", "baseline", "candidate"].includes(phase));
const git = (...args) =>
  execFileSync("git", args, { cwd: target, encoding: "utf8", timeout: 30_000 }).trim();
assert.equal(git("rev-parse", "HEAD"), spec.source);
assert.equal(git("rev-parse", "HEAD^{tree}"), spec.tree);
assert.equal(git("diff", "--name-only"), "", "Unstaged tracked source drift");
assert.equal(git("ls-files", "--others", "--exclude-standard"), "", "Unexpected untracked input");
const changed = phase === "original" ? {} : spec[phase];
assert.deepEqual(
  git("diff", "--cached", "--name-only", spec.source).split("\n").filter(Boolean).sort(),
  Object.keys(changed).sort(),
);
assert.equal(
  git("diff", "--cached", "--summary", spec.source),
  "",
  "Unexpected file mode or ownership change",
);
const expected = { ...spec.original, ...changed };
const hashes = {};
for (const [file, sha256] of Object.entries(expected)) {
  hashes[file] = createHash("sha256")
    .update(readFileSync(path.join(target, file)))
    .digest("hex");
  assert.equal(hashes[file], sha256, file);
}
console.log(
  JSON.stringify(
    {
      source: spec.source,
      tree: spec.tree,
      phase,
      changedPaths: Object.keys(changed).sort(),
      hashes,
    },
    null,
    2,
  ),
);

File: scripts/clawtributor-campaign50/lanes/slack-configured-ack/verify-source.py
import hashlib
import json
import pathlib
import subprocess
import sys

lane = pathlib.Path(sys.argv[1]).resolve()
phase = sys.argv[2]
assert phase in {"clean", "baseline", "candidate"}
source = json.loads((lane / "source.json").read_text())
manifest = json.loads((lane / "MANIFEST.json").read_text())

def run(*args):
    return subprocess.check_output(args, text=True).strip()

for name, expected in manifest["files"].items():
    assert hashlib.sha256((lane / name).read_bytes()).hexdigest() == expected, name
assert run("git", "rev-parse", "HEAD") == source["base"]
assert run("node", "--version") == "v" + source["node"]
assert run("pnpm", "--version") == source["pnpm"]
subprocess.run(["git", "diff", "--quiet"], check=True)
assert json.loads(pathlib.Path("package.json").read_text())["packageManager"] == source["packageManager"]
sdk_manifest = pathlib.Path(run("node", "-p", "require.resolve('@slack/web-api/package.json', {paths: [process.cwd() + '/extensions/slack']})"))
assert json.loads(sdk_manifest.read_text())["version"] == source["sdkVersion"]
sdk_hashes = {}
for name, expected in source["sdkFiles"].items():
    path = sdk_manifest.parent / "dist" / ("types/request/" + name if name.endswith(".d.ts") else name)
    sdk_hashes[name] = hashlib.sha256(path.read_bytes()).hexdigest()
    assert sdk_hashes[name] == expected, name
expected_files = dict(source["baseline"])
changed = []
if phase != "clean":
    expected_files[source["unit_path"]] = source["candidate"][source["unit_path"]]
    changed.append(source["unit_path"])
if phase == "candidate":
    expected_files.update(source["candidate"])
    changed = sorted(source["candidate"])
assert run("git", "diff", "HEAD", "--name-only").splitlines() == sorted(changed)
for name, expected in expected_files.items():
    assert hashlib.sha256(pathlib.Path(name).read_bytes()).hexdigest() == expected, name
untracked = run("git", "ls-files", "--others", "--exclude-standard").splitlines()
assert untracked == ([] if phase == "clean" else [source["test_path"]]), untracked
if phase != "clean":
    assert pathlib.Path(source["test_path"]).read_bytes() == (lane / "ack-ingress.proof.test.ts").read_bytes()
print(json.dumps({"phase": phase, "head": source["base"], "files": expected_files, "node": source["node"], "packageManager": source["packageManager"], "sdk": sdk_hashes}, indent=2))
